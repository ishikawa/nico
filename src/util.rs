use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
