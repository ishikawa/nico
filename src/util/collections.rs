use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
