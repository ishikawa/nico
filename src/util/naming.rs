use std::collections::HashMap;

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug, Default)]
pub struct SequenceNaming {
    names: HashMap<String, i32>,
}

impl SequenceNaming {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next<S: AsRef<str>>(&mut self, name: S) -> String {
        let name = name.as_ref();
        let n = self.names.entry(name.to_string()).or_insert(0);
        let next = format!("{}.{}", name, n);

        *n += 1;
        next
    }
}

/// This struct produces a unique name while giving a sequential
/// number to the name passed to the next() method.
#[derive(Debug)]
pub struct PrefixNaming {
    type_var_index: usize,
    prefix: String,
}

impl PrefixNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }

    pub fn next(&mut self) -> String {
        let i = self.type_var_index;
        let next = format!("{}{}", self.prefix, i);

        self.type_var_index += 1;
        next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefix_naming() {
        let mut naming = PrefixNaming::new("my_");

        assert_eq!(naming.next(), "my_0".to_string());
        assert_eq!(naming.next(), "my_1".to_string());
    }

    #[test]
    fn sequence_naming() {
        let mut naming = SequenceNaming::new();

        assert_eq!(naming.next("x"), "x.0".to_string());
        assert_eq!(naming.next("x"), "x.1".to_string());
        assert_eq!(naming.next("y"), "y.0".to_string());
    }
}
