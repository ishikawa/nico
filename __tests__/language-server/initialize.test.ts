import { LanguageServer, buildInitialize, buildInitialized } from "../util/lsp";

test("initialize", async done => {
  const server = LanguageServer.spawn();

  const request = buildInitialize();
  const response = await server.sendRequest(request);

  expect(response).toEqual({
    jsonrpc: "2.0",
    id: request.id,
    result: {
      capabilities: {
        semanticTokensProvider: {
          full: true,
          legend: {
            tokenModifiers: [
              "declaration",
              "definition",
              "readonly",
              "static",
              "deprecated",
              "abstract",
              "async",
              "modification",
              "documentation",
              "defaultLibrary"
            ],
            tokenTypes: [
              "keyword",
              "variable",
              "string",
              "number",
              "operator",
              "comment",
              "function",
              "struct",
              "function",
              "parameter",
              "property"
            ]
          }
        },
        textDocumentSync: 2
      },
      serverInfo: { name: "nico-ls", version: "0.0.1" }
    },
    error: null
  });

  // initialized
  {
    const notification = buildInitialized();
    await server.sendNotification(notification);
  }

  expect(server.isStopped()).toBeFalsy();

  server.on("exit", done);
  server.stop();
});
