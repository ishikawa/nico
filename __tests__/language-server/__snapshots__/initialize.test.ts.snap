// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialize 1`] = `
Object {
  "error": null,
  "id": 4760,
  "jsonrpc": "2.0",
  "result": Object {
    "capabilities": Object {
      "semanticTokensProvider": Object {
        "full": true,
        "legend": Object {
          "tokenModifiers": Array [
            "declaration",
            "definition",
            "readonly",
            "static",
            "deprecated",
            "abstract",
            "async",
            "modification",
            "documentation",
            "defaultLibrary",
          ],
          "tokenTypes": Array [
            "keyword",
            "variable",
            "string",
            "number",
            "operator",
            "comment",
            "struct",
            "function",
            "parameter",
            "property",
          ],
        },
      },
      "textDocumentSync": 2,
    },
    "serverInfo": Object {
      "name": "nico-ls",
      "version": "0.0.1",
    },
  },
}
`;

exports[`rename prepare support : off 1`] = `
Object {
  "error": null,
  "id": 4761,
  "jsonrpc": "2.0",
  "result": Object {
    "capabilities": Object {
      "renameProvider": true,
      "semanticTokensProvider": Object {
        "full": true,
        "legend": Object {
          "tokenModifiers": Array [
            "declaration",
            "definition",
            "readonly",
            "static",
            "deprecated",
            "abstract",
            "async",
            "modification",
            "documentation",
            "defaultLibrary",
          ],
          "tokenTypes": Array [
            "keyword",
            "variable",
            "string",
            "number",
            "operator",
            "comment",
            "struct",
            "function",
            "parameter",
            "property",
          ],
        },
      },
      "textDocumentSync": 2,
    },
    "serverInfo": Object {
      "name": "nico-ls",
      "version": "0.0.1",
    },
  },
}
`;

exports[`rename prepare support : on 1`] = `
Object {
  "error": null,
  "id": 4760,
  "jsonrpc": "2.0",
  "result": Object {
    "capabilities": Object {
      "renameProvider": Object {
        "prepareProvider": true,
      },
      "semanticTokensProvider": Object {
        "full": true,
        "legend": Object {
          "tokenModifiers": Array [
            "declaration",
            "definition",
            "readonly",
            "static",
            "deprecated",
            "abstract",
            "async",
            "modification",
            "documentation",
            "defaultLibrary",
          ],
          "tokenTypes": Array [
            "keyword",
            "variable",
            "string",
            "number",
            "operator",
            "comment",
            "struct",
            "function",
            "parameter",
            "property",
          ],
        },
      },
      "textDocumentSync": 2,
    },
    "serverInfo": Object {
      "name": "nico-ls",
      "version": "0.0.1",
    },
  },
}
`;
